# Rules for naming variable :

variable names can start with uuper and lowercase alphabet and also an underscore but not a no.

# Tpes of Variables:


int<br>
float<br>
complex<br>
bool<br>
string<br>
list<br>

In [6]:
a=1+2j
print(type(a))

<class 'complex'>


# TypeCasting:

Implicit typecasting:<Br> The variable automatically infers datatype of the variable no need to specify the datatype.<br>
Eg:

In [19]:
a=[10,20.31,4+3j]
print(list(map(type,a)))
print(type(a))

[<class 'int'>, <class 'float'>, <class 'complex'>]
<class 'list'>


Explicit Typecasting:<bR>
    Here we explicitly specify which data type the variable is.<br>
    Eg.

In [18]:
a=3.112
b=int(a)
print(type(b))

<class 'int'>


# How variables work in python:

In C++ or JAVA when a variable is specified with a value<br>
the variable name points to a memory location where the value is stored.<br>
but when value of the variable is changed then the value is actually updated at the same memory loaction .

However when a value is stored in a variable in python it points to a memory location as prior but when the value of the variable is changed then the variable now points to a new memory block and stores the new value there the prior value remains stored in the prior block only undisturbed by the changes made.

In [20]:
a=10
print(id(a))
a=20
print(id(a))

2016263105104
2016263105424


Thus even after being the same variable it made the change in new memory location.
So basically python variable points not to the memory location but the value specified.

In [21]:
c=20
print(id(c))

2016263105424


In [22]:
if id(a)==id(c):
    print(True)

True


So value 20 has the location '2016263105424' and not the variable a or c.

# Scope of a variable:

Two scopes:<BR>
    1. Local Scope: only valid till within a code block or indent block<br>
    2.Global Scope:for all the indent blocks present in the page

In [39]:
i=10
def scope_1():
    i=2
    print(f"inside function block i:{i}")
scope_1()
print(f"outside the function block i:{i}")

inside function block i:2
outside the function block i:10


if global variable is needed to be used in local scope then define it by global keyword.

In [40]:
def scope_2():
    global i
    print(i)
    i=2
    print(i)
scope_2()
print(i)

10
2
2


So the global i was set to 2 in the function.
